sklearn/ensemble/_hist_gradient_boosting/tests/test_binning.py

```diff
@@ -110,57 +112,60 @@ def test_map_to_bins(n_bins):
 def test_bin_mapper_random_data(n_bins):
     n_samples, n_features = DATA.shape
 
-    expected_count_per_bin = n_samples // n_bins
-    tol = int(0.05 * expected_count_per_bin)
-
     mapper = _BinMapper(max_bins=n_bins, random_state=42).fit(DATA)
     binned = mapper.transform(DATA)
 
     assert binned.shape == (n_samples, n_features)
     assert binned.dtype == np.uint8
-    assert_array_equal(binned.min(axis=0), np.array([0, 0]))
+    assert_array_equal(binned.min(axis=0), np.array([1, 1]))
     assert_array_equal(binned.max(axis=0), np.array([n_bins - 1, n_bins - 1]))
     assert len(mapper.bin_thresholds_) == n_features
     for bin_thresholds_feature in mapper.bin_thresholds_:
         assert bin_thresholds_feature.shape == (n_bins - 1,)
         assert bin_thresholds_feature.dtype == DATA.dtype
     assert np.all(mapper.actual_n_bins_ == n_bins)
 
-    # Check that the binned data is approximately balanced across bins.
+    # Check that the binned data is approximately balanced across bins
+    # (ignoring first bin since there are no missing values)
+    expected_count_per_bin = n_samples // (n_bins - 1)
+    tol = int(0.05 * expected_count_per_bin)
     for feature_idx in range(n_features):
-        for bin_idx in range(n_bins):
+        for bin_idx in range(1, n_bins):
             count = (binned[:, feature_idx] == bin_idx).sum()
             assert abs(count - expected_count_per_bin) < tol
 
 
-@pytest.mark.parametrize("n_samples, n_bins", [
+@pytest.mark.parametrize("n_samples, n_bins_for_non_missing", [
     (5, 5),
     (5, 10),
     (5, 11),
     (42, 255)
 ])
-def test_bin_mapper_small_random_data(n_samples, n_bins):
+def test_bin_mapper_small_random_data(n_samples, n_bins_for_non_missing):
     data = np.random.RandomState(42).normal(size=n_samples).reshape(-1, 1)
     assert len(np.unique(data)) == n_samples
 
-    mapper = _BinMapper(max_bins=n_bins, random_state=42)
+    max_bins = n_bins_for_non_missing + 1  # first bin reserved
+    mapper = _BinMapper(max_bins=max_bins, random_state=42)
     binned = mapper.fit_transform(data)
 
     assert binned.shape == data.shape
     assert binned.dtype == np.uint8
     assert_array_equal(binned.ravel()[np.argsort(data.ravel())],
-                       np.arange(n_samples))
+                       np.arange(n_samples) + 1)
 
 
-@pytest.mark.parametrize("n_bins, n_distinct, multiplier", [
+@pytest.mark.parametrize("n_bins_for_non_missing, n_distinct, multiplier", [
     (5, 5, 1),
     (5, 5, 3),
     (255, 12, 42),
 ])
-def test_bin_mapper_identity_repeated_values(n_bins, n_distinct, multiplier):
+def test_bin_mapper_identity_repeated_values(n_bins_for_non_missing,
+                                             n_distinct, multiplier):
     data = np.array(list(range(n_distinct)) * multiplier).reshape(-1, 1)
-    binned = _BinMapper(max_bins=n_bins).fit_transform(data)
-    assert_array_equal(data, binned)
+    max_bins = n_bins_for_non_missing + 1  # first bin reserved
```

Comment:
with all these it even more feels like having `max_bins` to include non-missing values makes more sense.